### PR TITLE
COMMERCE-9596 Extension point to define a Collection Provider that supplies SKU's as options

### PR DESCRIPTION
**Pull request by pietrobocale**
https://liferay.atlassian.net/browse/COMMERCE-9596
Base PR: https://github.com/liferay-commerce/liferay-portal/pull/4128

### CODE REVIEW

**Comment by pietrobocale** [comment]
ci:test:sf

**Comment by dsitu** [comment]
ci:test:relevant

**Comment by jaydawu** [comment]
ci:test:commerce-functional

**Comment by jaydawu** [comment]
ci:test:relevant

**Comment by jaydawu** [comment]
ci:test:commerce-functional

**Comment by dsitu** [comment]
ci:test:relevant

**Comment by jaydawu** [comment]
ci:test:relevant

**Comment by pietrobocale** [comment]
ci:stop

**Comment by pietrobocale** [comment]
ci:test:relevant

**Comment by pietrobocale** [comment]
ci:test:relevant

**Comment by dsitu** [comment]
ci:test:relevant

**Comment by dsitu** [comment]
Opened up a new PR [here](https://github.com/liferay-commerce/liferay-portal/pull/4139) to fix the conflicts.

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase [hunk 1]
```diff
@@ -0,0 +1,926 @@
...
+	}
+
+	tearDown {
+		CommerceConfiguration.commerceTearDown();
+	}
+
+	@description = "COMMERCE-12094. As a buyer user, I can view all the products with Define Externally configuration listed as option values in product details and checkout"
+	@priority = 5
+	test CanCheckoutBundledProductWithExternallyDefinedProductOptions {
+		property custom.properties = "jsonws.web.service.paths.excludes=${line.separator}company.security.strangers.verify=false";
```
Comment (lorenzolapescara) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase:19]: Please move this line at the definition level and remove all occurrences in the tests

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase [hunk 2]
```diff
@@ -0,0 +1,926 @@
...
+			CommerceProductOptions.editProductOption(
+				collectionProvider = "SKUs by Categories",
+				key_option = "Option Select from List",
+				key_priceType = "Dynamic",
+				productAddOptionCategoriesList = "Transmission",
+				productOptionDefineExternally = "true",
+				productOptionPriceType = "true");
+		}
+
+		task ("Then the buyer can view all the products related to that category added except U-Joint as options values in product details in the front store") {
```
Comment (lorenzolapescara) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase:70]: task ("Then the buyer can view all the products related to the category added as options values in the store front except U-Joint") {

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase [hunk 3]
```diff
@@ -0,0 +1,926 @@
...
+				requireReset = "false",
+				userEmailAddress = "buyer@liferay.com",
+				userFirstName = "Buyer",
+				userLastName = "User",
+				userRole = "Buyer",
+				userScreenName = "buyeruser",
+				userSiteMembership = "Minium");
+		}
+
+		task ("And a 50% discount that is applied to the U-joint product") {
```
Comment (lorenzolapescara) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase:146]: task ("And a 50% discount is applied to the U-joint product") {

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase [hunk 4]
```diff
@@ -0,0 +1,926 @@
...
+
+			AssertTextEquals(
+				locator1 = "CommerceAccelerators#CHECKOUT_SUCCESS_MESSAGE",
+				value1 = "Success! Your order has been processed.");
+		}
+	}
+
+	@description = "COMMERCE-12111. As a buyer user, I can checkout discount products using Define Externally configuration"
+	@priority = 4
+	test CanDiscountWorkWithDefineExternallyProducts {
```
Comment (lorenzolapescara) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase:127]: CanApplyDiscountToExternallyDefinedProducts

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase [hunk 5]
```diff
@@ -0,0 +1,926 @@
...
+				productName = "U-Joint");
+		}
+
+		task ("And the admin opens the U-Joint Product via Admin") {
+			CommerceProducts.openProductsAdmin();
+
+			CommerceProducts.searchAndGoToProductEntry(entryProductName = "U-Joint");
+		}
+
+		task ("And a new option with three values is created") {
```
Comment (lorenzolapescara) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase:166]: task ("And a new option is created") {

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase [hunk 6]
```diff
@@ -0,0 +1,926 @@
...
+			CommerceProductOptions.editProductOption(
+				collectionProvider = "SKUs by Categories",
+				key_option = "Option Select from List",
+				key_priceType = "Dynamic",
+				productAddOptionCategoriesList = "Transmission",
+				productOptionDefineExternally = "true",
+				productOptionPriceType = "true");
+		}
+
+		task ("And the buyer adds a product Calipers as an option value to the cart in U-Joint product details") {
```
Comment (lorenzolapescara) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase:193]: task ("And the buyer adds Calipers as an option value to the cart in U-Joint product details") {

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase [hunk 7]
```diff
@@ -0,0 +1,926 @@
...
+					addressName = "Address Name",
+					deliveryValue = 15,
+					productDiscountedTotalPrice = "13.5",
+					productDiscountValue = "13.5",
+					productName = ${productName},
+					productPrice = 27,
+					productQuantity = 1,
+					subtotalValue = 15,
+					totalValue = 30);
+			}
```
Comment (lorenzolapescara) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase:258]: No need for the for cycle here

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase [hunk 8]
```diff
@@ -0,0 +1,926 @@
...
+
+			AssertElementNotPresent(
+				key_entryName = "ABS Sensor",
+				locator1 = "CommerceEntry#TABLE_ELIGIBILITY_ACCOUNT_GROUPS");
+
+			echo("The product ABS Sensor is not present anymore");
+		}
+	}
+
+	@description = "COMMERCE-12137 Verify that is possible only to use the Dynamic Price Type"
```
Comment (lorenzolapescara) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase:516]: Add some details in the description

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase [hunk 9]
```diff
@@ -0,0 +1,926 @@
...
+	@priority = 4
+	test CanUseOnlyDynamicPriceType {
```
Comment (lorenzolapescara) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase:518]: CanUseDefineExternallyOnlyWithDynamicPriceType

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase [hunk 10]
```diff
@@ -0,0 +1,926 @@
...
+			AssertElementPresent(
+				key_title = "Price Type",
+				locator1 = "CommerceEntry#CATALOGS_FORM",
+				value1 = "");
+		}
+	}
+
+	@description = "COMMERCE-12093. As an admin, I can create a bundled product with options using Define Externally configuration"
+	@priority = 5
+	test CanViewExternallyDefinedProductsListedAccordingToSelectedCategory {
```
Comment (lorenzolapescara) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase:604]: Please adjust description and test name to fit the steps executed, thanks!

File: modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase [hunk 11]
```diff
@@ -0,0 +1,926 @@
...
+			}
+
+			AssertElementNotPresent(
+				key_entryName = "U-Joint",
+				locator1 = "CommerceEntry#TABLE_ELIGIBILITY_ACCOUNT_GROUPS");
+
+			echo("The product U-Joint is not present");
+		}
+
+		task ("And the buyer can view all the products related to that category are added except U-Joint as options values in product details in the front store") {
```
Comment (lorenzolapescara) [inline modules/apps/commerce/commerce-product-test/src/testFunctional/tests/CPCommerceProductDefineExternalyOption.testcase:813]: task ("And the buyer can assert that all the products related to that category are added except U-Joint as options values in product details in the front store") {

**Review by lorenzolapescara — COMMENTED** [review]
Review completed @pietrobocale

**Pull request closed without merging**